src/expr/node_algorithm.h

```diff
@@ -0,0 +1,130 @@
+/*********************                                                        */
+/*! \file node_algorithm.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Morgan Deters, Andrew Reynolds, Tim King
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2018 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Common algorithms on nodes
+ **
+ ** This file implements common algorithms applied to nodes, such as checking if
+ ** a node contains a free or a bound variable. This file should generally only
+ ** be included in source files.
+ **/
+
+#include "cvc4_private.h"
+
+#ifndef __CVC4__EXPR__NODE_ALGORITHM_H
+#define __CVC4__EXPR__NODE_ALGORITHM_H
+
+#include <unordered_map>
+#include <vector>
+
+#include "expr/node.h"
+
+namespace CVC4 {
+namespace expr {
+
+/**
+ * A class that pretends iterating over a node of a certain kind. If the node
+ * matches the kind, the class iterates over the children, otherwise the
+ * iterator just points to the node. This is useful if you want to pretend to
+ * iterate over a "unary" PLUS, for instance, since unary PLUSes don't exist.
+ */
+class KindedIterator
```

Comment:
I suggest keeping KindedIterator unchanged & unmoved in this PR.

Instead go forward with the two algorithms, which have plenty of usages, and send a different PR to remove KindedIterator entirely. It appears to have exactly 1 usage. Just simulate what this class is providing by doing this there. Unless you are planning on adding a lot of future usages, in which case add it to its own file.

Another option for this PR is to move this into src/prop/prop_engine.cpp.

Reply:
I suggest to remove the kinded_iterator and change the loop in `src/prop/prop_engine.cpp`. If the iterator was not really used until now, I doubt there will be more future usages.

Reply:
I've actually come across other use cases. Here's [one example](https://github.com/CVC4/CVC4/blob/master/src/theory/strings/theory_strings_rewriter.cpp#L2072-L2097). I think the main reason for not using it was that people weren't aware of its existence (at least that's what the reason was for me when I wrote this code). I do agree that we can exclude it from this PR, however.

Reply:
We could put it in separate files (e.g. `node_iterator.{h,cpp}`) where we collect helpers for iterating over nodes (e.g. an iterator that traverses in depth-first order and an iterator that flattens while traversing).

Reply:
I have removed the changes to the kinded iterator from the changeset.